Chess board test improvements

frozenfrank opened the pull request:
## Key Points

- **This PR does not change any course requirements.** The `ChessBoard` class has always required `equals()` and `hashCode()` testing; however, these tests were previously not explicit. Students would fail without good implementation, but the symptoms were unclear.
- This PR only adds additional tests which better evaluate existing requirements.
- These new tests have been tested against functioning code.

## Overview

Improve comprehensiveness and make ChessBoard tests more understandable by adding specific tests for requirements that students would otherwise discover late.

## Discussion

It oftentimes takes students a long time to realize that a `new ChessBoard()` cannot call `resetBoard()` automatically. While this isn't the place to explain _why_ this is the case, we can point out the requirement explicitly.

This test uses some randomness while testing the hashCodes of `ChessBoard`. This randomness reduces the total number of evaluations required, but still emphasizes the overall requirement of the test. I think it was fun to create the tests.

## Dependencies

- ~This is intended to be merged **after** #16.~
    - Actually, it doesn't matter much which order these PRs are merged.
    - It would be better to merge #16 first since then both merge commits end up on the same branch, but it will be fine either way.
    - I mostly just wanted to raise these issues separate from the improvements provided in the base PR.

## Work List
- [x] Run working implementation through revamped test cases

## Help Requested

@Maillman Can you please run your code through these tests? This branch also includes the changes in #16.

leesjensen reviewed (approved):
Thanks!

frozenfrank committed:
Provide meaningful different boards for hashCode testing

frozenfrank committed:
Add 'Construct Empty ChessBoard' test

AldenKim commented:
James, I am having some weird behavior with the new tests. In the chess board tests, the Equals & hashcode test, the hashcode test, and the equals test sometimes pass, sometimes fail. Do you think it could be my code? This could be unintended behavior from the tests as well. What do you think?
![Screenshot 2025-01-28 130449](https://github.com/user-attachments/assets/e735d367-a141-419e-8422-e0977033fc0c)
![Screenshot 2025-01-28 130502](https://github.com/user-attachments/assets/70ef18a5-0339-48d7-8fa0-47744f364dec)

ZakkeryDaRebel commented:
I also tried to run the code and have the exact same test failing for me. Which means that Alden and I both have the same problem in our code, or something in the test is off.

frozenfrank commented:
> I also tried to run the code and have the exact same test failing for me. Which means that Alden and I both have the same problem in our code, or something in the test is off.

This is why we test it before we thrust all these problems on the students. Thanks for running your tests through it. We'll need to get to the bottom of it before we release these.

frozenfrank committed:
Fix issue with failing ChessBoard tests...

The root cause stemmed from the probability that a piece
would be replaced with the exact same piece. When this happens,
the board is not actually different, and test would fail improerly.

Since there are 6 pieces * 2 colors = 12 pieces, the probabiliy
of an individual piece being replaced with itself is 1 in 12.
This is about 9%. However, there are 14 different pieces that
each have this probability. The total probability compounds
up to ~66%!

Because of this overlooked error, the tests would fail improperly
most of the time.

This change does not call resetBoard() before creating each of the boards.
Therefore, they are all blank, and adding a random piece in a different
position each time always creates a definitely different board.

frozenfrank commented:
@AldenKim @ZakkeryDaRebel  Wow, thanks for identifying this issue! I implemented `ChessBoard` myself again (for fun), and also saw this behavior.

After thinking about it for a bit, the root problem is that I was replacing a square on the board and expecting it to always be different, but sometimes I "replaced" the piece with the very same piece. The probabilities of error compounded due to the thoroughness of the test, and the test only worked with probability $\left(1-\frac{1}{12}\right)^{14} \approx \text{29.6} \\%$. Ouch.

This issue has been rectified and described in more detail in commit 7d2cc8e.

<img width="632" alt="image" src="https://github.com/user-attachments/assets/9a2120ea-1ad7-4333-8bca-d37c734987d1" />

frozenfrank committed:
Reorder ChessBoard tests to read in an order students are more likely to implement them

frozenfrank commented:
With all these tests, we are now sitting at 67 tests.

<img width="605" alt="image" src="https://github.com/user-attachments/assets/f4fb1082-8c79-4f74-a045-cf208ae4ec8f" />

ZakkeryDaRebel commented:
Hmm, maybe it's just my code, but now I seem to be failing 1-3 tests (it changes each time I run it), they are the ChessBoardTests' Equals & HashCode, HashCode, and Equals Testing. But I have IntelliJ generate the equals and hashcode for me, so I'm uncertain of the issue. 
<img width="1728" alt="Screenshot 2025-01-29 at 1 22 56 PM" src="https://github.com/user-attachments/assets/b9678315-c173-4bcc-b120-80b23494412c" />
<img width="1728" alt="Screenshot 2025-01-29 at 1 26 53 PM" src="https://github.com/user-attachments/assets/061a9d51-a77d-4f98-8b8c-b11cc57424b4" />

frozenfrank commented:
> Hmm, maybe it's just my code, but now I seem to be failing 1-3 tests (it changes each time I run it), they are the ChessBoardTests' Equals & HashCode, HashCode, and Equals Testing. But I have IntelliJ generate the equals and hashcode for me, so I'm uncertain of the issue.

@ZakkeryDaRebel The screenshots are very helpful, but they don't reveal exactly which commits you have. Did you update your branch since I pushed a few fixes? Can you show the `git log` output (the fix is contained in commit 7d2cc8e)?

AldenKim commented:
> > Hmm, maybe it's just my code, but now I seem to be failing 1-3 tests (it changes each time I run it), they are the ChessBoardTests' Equals & HashCode, HashCode, and Equals Testing. But I have IntelliJ generate the equals and hashcode for me, so I'm uncertain of the issue.
> 
> @ZakkeryDaRebel The screenshots are very helpful, but they don't reveal exactly which commits you have. Did you update your branch since I pushed a few fixes? Can you show the `git log` output (the fix is contained in commit [7d2cc8e](https://github.com/softwareconstruction240/chess/commit/7d2cc8e68b283b9b10d451d6e3a39b83b3ee487c))?

The new tests are working consistently on my end now. Are you sure you have the most recent commits?
![Screenshot 2025-01-29 133337](https://github.com/user-attachments/assets/25d194ba-b16e-47c9-a103-90f4355a679e)

ZakkeryDaRebel commented:
Yeah, that's my bad, I thought I had updated my branch, but I guess I didn't. Now that I have updated it, it all looks good, all 64 of the tests are working. Sorry about that.

frozenfrank committed:
Remove randomness and pawns from back row

frozenfrank committed:
Merge branch 'chess-position-tests' into chess-board-test-improvements

frozenfrank committed:
Fix array index issue and improve readability

frozenfrank commented:
@softwareconstruction240/professors This PR is now ready for review :)

A few notes about the current PR strategy:
* There are several smaller PRs. This allows for:
    * More granular investigation into changes
    * Independent approval/rejection of individual ideas
* This is the first PR with a merge dependency
    * I'm requesting review for items in _one_ order they could be approved
    * Any merge dependencies are clearly listed in the PR description
    * The dependent PRs are shown in an expanded format which allows the reader to easily determine if it has already been merged.
* So far, none of the PRs have required changes to the course requirements
* A new set of tags have been created to identify how each PR interacts with course requirements.
    
![image](https://github.com/user-attachments/assets/ed0ea9ea-23ad-4f56-b060-757314a00346)
